src/EditorFeatures/Test/Rename/RenameRewriterTests.cs

```diff
@@ -0,0 +1,497 @@
+
+            await verifier.RenameAndAnnotatedDocumentAsync(
+                documentFilePath: "test1.cs",
+                new()
+                {
+                    { "Rename1", ("Apple2", renameOption) },
+                    { "Rename2", ("Lemon2", renameOption) },
+                    { "Rename3", ("banana2", renameOption) },
+                }, cancellationToken);
+            await verifier.VerifyDocumentAsync("test1.cs",
+@"
+/// <summary>
+/// <see cref=""Apple2""/> Apple2 is not Lemon2. Lemon2 is not banana2
+/// </summary>
+/// <Lemon2>
+/// </Lemon2>
+public class Apple2
+{
+    // banana2 is not Apple2
+    public void Lemon2(int banana2)
+    {
+        string Apple = ""Apple2, Lemon2 and banana2 are fruit"";
+        string Lemon = $""Apple2, Lemon2 and {banana2} are fruit"";
+    }
+}", cancellationToken);
+        }
+
+        [Fact]
+        public async Task TestCSharpRenameComplexification()
+        {
+            var cancellationToken = CancellationToken.None;
+            using var verifier = new Verifier(@"
+                   <Workspace>
+                       <Project Language=""C#"" CommonReferences=""true"">
+                           <Document FilePath=""test1.cs"">
+<![CDATA[
+public class {|Rename1:Apple|}
+{
+    public void Lemon(int banana)
+    {
+        {|Conflict:{|Rename2:banana2|}|} = 10;
+    }
+    public static int banana2
+    {
+        get; private set;
+    }
+}]]>
+                           </Document>
+                    </Project>
+                </Workspace>
+");
+
+            var renameOption = new SymbolRenameOptions();
+            await verifier.RenameAndAnnotatedDocumentAsync(
+                documentFilePath: "test1.cs",
+                new()
+                {
+                    { "Rename1", ("Apple2", renameOption) },
+                    { "Rename2", ("banana", renameOption) },
+                }, cancellationToken);
+            await verifier.VerifyDocumentAsync("test1.cs",
+@"
+public class Apple2
+{
+    public void Lemon(int banana)
+    {
+        global::Apple2.banana = 10;
+    }
+    public static int banana
+    {
+        get; private set;
+    }
+}", cancellationToken);
+        }
+
+        [Fact]
+        public async Task TestRenameFailed()
+        {
+            var cancellationToken = CancellationToken.None;
+            using var verifier = new Verifier(@"
+                   <Workspace>
+                       <Project Language=""C#"" CommonReferences=""true"">
+                           <Document FilePath=""test1.cs"">
+interface IBar
+{
+    void {|Rename1:Hello|}();
+}
+
+class Bar : IBar
+{
+    public void {|Rename2:Hello|}();
+}
+
+                           </Document>
+                    </Project>
+                </Workspace>
+");
+
+            var renameOption = new SymbolRenameOptions();
+            await Assert.ThrowsAsync<ArgumentException>(() =>
```

Comment:
It is thrown from [place1](https://github.com/Cosifne/roslyn/blob/dd0838742da0dcda1cc820912b04fa9c97d3fd1e/src/Workspaces/Core/Portable/Rename/IRenameRewriterLanguageService.cs#L290)
and [place2](https://github.com/Cosifne/roslyn/blob/dd0838742da0dcda1cc820912b04fa9c97d3fd1e/src/Workspaces/Core/Portable/Rename/IRenameRewriterLanguageService.cs#L187)

It assume that for a text span of a syntax tree, we should only try to rename the underlying token to one replacement text.
This won't be a problem for the single symbol case, (i.e. the references to a single symbol won't overlapped, otherwise it would be a problem somewehre in the FAR engine)

But this might be a problem for two symbols cases (as shown in the tests)

Reply:
What I want is an error-checking logic in the rename rewriter. so how about create a renamer specific exception?

Reply:
I'd say do one of two things:

1. Make a new exception type and assert that.
2. Look at the exception returned from ThrowsAsync() and assert the text of the message matches the failure you're expecting to see.

The former is a bit nicer in terms of type safety -- you could put the failure information as properties and then assert those as well. The latter is perhaps easier, but still guarantees the same thing with less work. The former might also be better than putting user info into the exception message too which might make it easier when looking at any fault reporting that may happen, since if you are looking at telemetry and do a group by fault message, you'll end up with a bunch of individual messages.

Reply:
I think this conversation is related to the https://github.com/dotnet/roslyn/pull/62670#discussion_r937170083 below. 
Let's see if we would just annotate it as a conflict or throw an exception.

Reply:
I have added specific exception types